### PR TITLE
Support varlen FlashAttention

### PR DESCRIPTION
**Pull request by Seventeen17**

### CODE REVIEW

File: torchacc/ops/flash_attn.py [hunk 1]
```diff
@@ -165,24 +165,23 @@ def backward(ctx, dout, *args):
...
-    def forward(ctx, q, k, v, cu_seqlens_q, cu_seqlens_k, max_seqlen_q,
-                max_seqlen_k, dropout_p, softmax_scale, causal, window_size,
-                alibi_slopes, deterministic, return_softmax):
+    def forward(ctx, q, k, v, attention_mask, dropout_p, softmax_scale, causal,
+                window_size, alibi_slopes, deterministic, return_softmax):
         if softmax_scale is None:
             softmax_scale = q.shape[-1]**(-0.5)
         assert isinstance(window_size, tuple) and len(window_size) == 2
 
-        softmax_lse, out, rng_state = torch_xla._XLAC._flash_attention_forward(
```
Comment (baoleai) [inline torchacc/ops/flash_attn.py:175]: `SPMDFlashAttnVarlenXla` and `FlashAttnVarlenQKVPackedXla` also need to be updated.

Comment (Seventeen17) [reply torchacc/ops/flash_attn.py:175]: Done. But I found no UT for them.

File: tests/ops/test_flash_attn.py [hunk 2]
```diff
@@ -30,42 +30,34 @@ def setup_env():
...
-# seqlen_q = 113, seqlen_k = 203, d = 32, dropout_p = 0.0, causal = True, local = False, alibi = False, deterministic = False, mha_type = 'mha', dtype = torch.float16
-# seqlen_q = 113, seqlen_k = 203, d = 256, dropout_p = 0.0, causal = False, local = False, alibi = True, deterministic = False, mha_type = 'mqa', dtype = torch.bfloat16
-# seqlen_q = 512, seqlen_k = 256, d = 32, dropout_p = 0.0, causal = False, local = False, alibi = True, deterministic = False, mha_type = 'gqa', dtype = torch.bfloat16
-# seqlen_q = 113, seqlen_k = 203, d = 32, dropout_p = 0.17, causal = False, local = False, alibi = False, deterministic = False, mha_type = 'mha', dtype = torch.float16
-
-
-@pytest.mark.parametrize("dtype", [torch.float16, torch.bfloat16])
-@pytest.mark.parametrize("mha_type", ["mha", "mqa", "gqa"])
-@pytest.mark.parametrize("deterministic", [False, True])
+@pytest.mark.parametrize("dtype", [torch.bfloat16])
```
Comment (anw90) [inline tests/ops/test_flash_attn.py]: Why are we removing so many test options?

File: tests/ops/test_flash_attn_varlen.py [hunk 3]
```diff
@@ -0,0 +1,279 @@
...
+    cu_seqlens = F.pad(
+        torch.cumsum(seqlens_in_batch, dim=0, dtype=torch.int32), (1, 0))
+    return (
+        indices,
+        cu_seqlens,
+        max_seqlen_in_batch,
+    )
+
+
+class FlashAttention2(nn.Module):
```
Comment (anw90) [inline tests/ops/test_flash_attn_varlen.py]: What does the FlashAttention2 mean? Does FA2 stand for it, or is the 2 just a magic number?

File: tests/ops/test_flash_attn_varlen.py [hunk 4]
```diff
@@ -0,0 +1,279 @@
...
+
+    with torch.cuda.amp.autocast(dtype=dtype):
+        ret_xla = model_xla(q_xla, k_xla, v_xla, attention_mask=None)
+
+    (
+        dq_xla,
+        dk_xla,
+        dv_xla,
+    ) = torch.autograd.grad(ret_xla, (q_xla, k_xla, v_xla), g_xla)
+    ta.mark_step()
```
Comment (anw90) [inline tests/ops/test_flash_attn_varlen.py]: use ta.sync instead.

File: tests/ops/test_flash_attn.py (hunk 2 above)
Comment (Seventeen17) [reply tests/ops/test_flash_attn.py]: added back

File: tests/ops/test_flash_attn_varlen.py (hunk 3 above)
Comment (Seventeen17) [reply tests/ops/test_flash_attn_varlen.py]: Just a magic number, removed

File: tests/ops/test_flash_attn_varlen.py (hunk 4 above)
Comment (Seventeen17) [reply tests/ops/test_flash_attn_varlen.py]: done

**Review by anw90 — APPROVED** [review]

**Pull request merged**